airbyte_cdk/sources/declarative/requesters/query_properties/query_properties.py

```diff
@@ -0,0 +1,51 @@
+# Copyright (c) 2025 Airbyte, Inc., all rights reserved.
+
+from dataclasses import InitVar, dataclass
+from typing import Any, Iterable, List, Mapping, Optional, Union
+
+from airbyte_cdk.sources.declarative.requesters.query_properties import (
+    PropertiesFromEndpoint,
+    PropertyChunking,
+)
+from airbyte_cdk.sources.types import Config, StreamSlice
+
+
+@dataclass
+class QueryProperties:
+    """
+    Low-code component that encompasses the behavior to inject additional property values into the outbound API
+    requests. Property values can be defined statically within the manifest or dynamically by making requests
+    to a partner API to retrieve the properties. Query properties also allow for splitting of the total set of
+    properties into smaller chunks to satisfy API restrictions around the total amount of data retrieved
+    """
+
+    property_list: Optional[Union[List[str], PropertiesFromEndpoint]]
+    always_include_properties: Optional[List[str]]
+    property_chunking: Optional[PropertyChunking]
+    config: Config
+    parameters: InitVar[Mapping[str, Any]]
+
+    def get_request_property_chunks(
+        self, stream_slice: Optional[StreamSlice] = None
+    ) -> Iterable[List[str]]:
+        fields: Union[Iterable[str], List[str]]
+        if isinstance(self.property_list, PropertiesFromEndpoint):
+            fields = self.property_list.get_properties_from_endpoint(stream_slice=stream_slice)
+        else:
+            fields = self.property_list if self.property_list else []
+
+        if self.property_chunking:
+            yield from self.property_chunking.get_request_property_chunks(
+                property_fields=fields, always_include_properties=self.always_include_properties
+            )
+        else:
+            yield from [list(fields)]
+
+    def has_multiple_chunks(self, stream_slice: Optional[StreamSlice]) -> bool:
```

Comment:
nit: It feels like we could let that to the consumer of the query properties (in our case, SimpleRetriever) we don't need to maintain this new method. Do we know in which other case we would need this method? I'm mostly worried about maintaining multiple methods that might generate multiple same HTTP requests multiple times

Reply:
I think that's fair and it does give the impression this is a "cheap" method, but in reality makes an API call. The main reason i ended up adding it was that we needed to introspect the iterable to detect if there were at least two elements. And we've been burned in the past with `tee()` not deallocating slices properly.

But I think we can move this method up into the retriever and if we combine with caching `get_request_property_chunks()` that should make things less require fewer outbound API calls. Thanks for the suggestion

Reply:
> we needed to introspect the iterable to detect if there were at least two elements

We could also increment a variable to know how many chunks were generated and take action on this. I don't have a preference though

Reply:
I thought about that too, but one problem with that is that is we want to know whether we need to start accumulating records in the merge dictionary on the first chunk, so the incrementing variable only allows us to know on the second loop. So that was why the alternative was a `tee()`. 

Alternatively, I don't love it, but I would also expect the property list to generally not be that big, so we could convert it to a list and just get the length. but I still lean towards just using this method twice since it should be cached

Reply:
i think that makes sense, I tweaked the design a bit to be agnostic to how many chunks. And also made a few adjustments so that even if we don't get a merge key, we still ensure that we emit the record